test-codegen: replace tempname/open/close/stat by mktemp

tkluck opened the pull request:
This seems like a [flaky test on Windows][1],
```
    FAILURE
Error in testset codegen:
Test Failed
  Expression: tempty == true
   Evaluated: false == true
ERROR: LoadError: Test run finished with errors
while loading C:\projects\julia\julia-a661736be5\share\julia\test\runtests.jl, in expression starting on line 29
Command exited with code 1
```
and without being able to reproduce, I'm going to guess it's a race condition betweeen `close` and `stat`. If so, replacing the `stat (2)` syscall by `fstat (2)` may fix it.  At the Julia level, that means calling `stat` on the iostream instead.

[1]: https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.17932/job/kn10okpwqrbs0126

tkluck committed:
test-codegen: replace tempname/open/close/stat by mktemp

This seems like a flaky test on Windows [1],

        FAILURE
    Error in testset codegen:
    Test Failed
      Expression: tempty == true
       Evaluated: false == true
    ERROR: LoadError: Test run finished with errors
    while loading C:\projects\julia\julia-a661736be5\share\julia\test\runtests.jl, in expression starting on line 29
    Command exited with code 1

and without being able to reproduce, I'm going to guess it's a race
condition betweeen `close` and `stat`. If so, replacing the `stat (2)`
syscall by `fstat (2)` may fix it.  At the Julia level, that means
calling `stat` on the iostream instead.

[1]: https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.17932/job/kn10okpwqrbs0126

tkelman commented on test/codegen.jl:
```diff
@@ -49,36 +49,32 @@ end
+        test_jl_dump_compiles_internal(1)
+        ccall(:jl_dump_compiles, Void, (Ptr{Void},), C_NULL)
+
+        @test stat(io).size == 0
```
isn't this changing the sense of the test?

tkluck replied:
Hmm you think? Maybe I'm missing something, but it seemed to me the intention is to test what gets written to the `io` handle that we pass to `:jl_dump_compiles` -- namely, test that nothing gets written (and/or (unlikely but possible) that everything gets truncated after writing). In any case, shouldn't that be the same between calling `fstat` on the file handle and calling `stat` on the file name?

nalimilan replied:
I think what @tkelman meant is that `tempty == false` was equivalent to `tstats.size != 0` in the existing code. Now you made it test that this is `== 0`.

tkluck replied:
Oops! I missed that the two test cases were different. Thanks for pointing that out, @nalimilan and @tkelman . In addition, the reason why the test case still passed is that `fstat` only gives the correct size after `flush`.

Let's see what AppVeyor says.

tkluck committed:
test-codegen: flush() before fstat()

This fixes two issues:

1. I replaced an `tempty == false` by `size == 0` (should have been `size != 0`)
2. I didn't notice that because I didn't `flush()` before `stat`ing the handle.

The second point is ironic, since that's exactly the kind of issue this
branch is trying to avoid! See commit message for d7e52d8bfff924 for
details.

fredrikekre commented:
Think this happened on Travis too https://gist.github.com/fredrikekre/ee13a0a48230346343da48e721c0b5c4
```
	From worker 1:		From worker 1:	Worker 10 failed running test codegen:
Some tests did not pass: 86 passed, 1 failed, 0 errored, 0 broken.codegen: Test Failed
  Expression: tempty == true
   Evaluated: false == true
Stacktrace:
 [1] record(::Base.Test.DefaultTestSet, ::Base.Test.Fail) at ./test.jl:651
 [2] (::##40#44)() at /private/tmp/julia/share/julia/test/runtests.jl:160
 [3] cd(::##40#44, ::String) at ./file.jl:70
 [4] include_relative(::Module, ::String) at ./loading.jl:464
 [5] include(::Module, ::String) at ./sysimg.jl:14
 [6] process_options(::Base.JLOptions) at ./client.jl:315
 [7] _start() at ./client.jl:383
```

fredrikekre commented:
Subsumed by #23486 ?

tkluck commented:
@fredrikekre looks like it! Closing.
